src/platform/backends/hyperv_api/hcn/hyperv_hcn_api_wrapper.cpp

```diff
@@ -0,0 +1,353 @@
+
+    const auto input = [&guid_wstr]() {
+        switch (guid_wstr.length())
+        {
+        case kGUIDLength:
+            // CLSIDFromString requires GUIDs to be wrapped with braces.
+            return fmt::format(L"{{{}}}", guid_wstr);
+        case kGUIDLengthWithBraces:
+        {
+            if (guid_wstr.front() != L'{' || guid_wstr.back() != L'}')
+            {
+                throw GuidParseError{"GUID string either does not start or end with a brace."};
+            }
+            return guid_wstr;
+        }
+        }
+        throw GuidParseError{"Invalid length for a GUID string ({}).", guid_wstr.length()};
+    }();
+
+    ::GUID guid = {};
+
+    const auto result = CLSIDFromString(input.c_str(), &guid);
+
+    if (FAILED(result))
+    {
+        throw GuidParseError{"Failed to parse the GUID string ({}).", result};
+    }
+
+    return guid;
+}
+
+/**
+ * Parse given GUID string into a GUID struct.
+ *
+ * @param guid_str GUID in string form, either 36 characters
+ *                  (without braces) or 38 characters (with braces.)
+ *
+ * @return GUID The parsed GUID
+ */
+auto guid_from_string(const std::string& guid_wstr) -> ::GUID
+{
+    const std::wstring v = maybe_widen{guid_wstr};
+    return guid_from_string(v);
+}
+
+// ---------------------------------------------------------
+
+/**
+ * Perform a Host Compute Network API operation
+ *
+ * @param fn The API function pointer
+ * @param args The arguments to the function
+ *
+ * @return HCNOperationResult Result of the performed operation
+ */
+template <typename FnType, typename... Args>
+OperationResult perform_hcn_operation(const HCNAPITable& api, const FnType& fn, Args&&... args)
+{
+    // Ensure that function to call is set.
+    if (nullptr == fn)
+    {
+        assert(0);
+        // E_POINTER means "invalid pointer", seems to be appropriate.
+        return {E_POINTER, L"Operation function is unbound!"};
+    }
+
+    // HCN functions will use CoTaskMemAlloc to allocate the error message buffer
+    // so use UniqueCotaskmemString to auto-release it with appropriate free
+    // function.
+    UniqueCotaskmemString result_msgbuf{};
+
+    // Perform the operation. The last argument of the all HCN operations (except
+    // HcnClose*) is ErrorRecord, which is a JSON-formatted document emitted by
+    // the API describing the error happened. Therefore, we can streamline all API
+    // calls through perform_hcn_operation.
+    const auto result =
+        ResultCode{fn(std::forward<Args>(args)..., out_ptr(result_msgbuf, api.CoTaskMemFree))};
+
+    mpl::debug(kLogCategory, "perform_hcn_operation(...) > result: {}", static_cast<bool>(result));
+
+    // Error message is only valid when the operation resulted in an error.
+    // Passing a nullptr is well-defined in "< C++23", but it's going to be
+    // forbidden afterwards. Going an extra mile just to be future-proof.
+    return {result, {result_msgbuf ? result_msgbuf.get() : L""}};
```

Comment:
Here's a suggestion to make this less verbose to the reader:

```suggestion
    // avoid null to be forward-compatible with C++23
    return {result, {result_msgbuf ? result_msgbuf.get() : L""}};
```

Reply:
done